readme.md
# beauty-logger

[ENGLISH] | [中文](https://github.com/zhoushoujian/beauty-logger/blob/master/readme_zh.md)

`A beautiful logger system for both nodejs and browser`

## Tip

For browser logger, it's better to use [logger-for-cannot-duplicate](https://github.com/zhoushoujian/logger-for-cannot-duplicate)

## Usage

```js
const path = require('path');
const Logger = require('beauty-logger');

const logger1 = new Logger({
  logFilePath: path.join(__dirname, './server.log'),
});

const logger2 = new Logger({
  //max size of per log file, default: 10MB
  logFileSize: 1024 * 1024 * 5,
  logFilePath: {
    //log file name, default: as follows
    info: path.join(__dirname, './info.log'),
    warn: path.join(__dirname, './warn.log'),
    error: path.join(__dirname, './error.log'),
  },
  //enable data type warn, default: false
  dataTypeWarn: true,
  //disable print log in console, default: false
  productionModel: false,
  //only print log in console, default: false
  onlyPrintInConsole: false,
  //execute other beauty logger instance at the same time，default：[]
  otherBeautyLoggerInstances: [],
  //execute callback when logging, default: null
  //level: log level
  //data: log content
  //pid: process pid
  //filePath: execute log function filename
  //content: pure log content, without time and pid and etc.
  callback: (level, data, pid, filePath， content) => void,
  //count package usage to upload
  uploadPackageInfoUrl: ""
});
logger1.info('logger1', 'beauty-logger'); // [2021-12-21 16:30:18.998] [INFO] [charms-Mac-Pro.local] [3400] [lib/beauty-logger.ts:323] "logger1"   [ext] "beauty-logger"
logger2.info('logger2', 'beauty-logger'); // [2021-12-21 16:30:18.998] [INFO] [charms-Mac-Pro.local] [3400] [lib/beauty-logger.ts:323] "logger2"   [ext] "beauty-logger"
```

## Apis

1. logger.debug
2. logger.info
3. logger.warn
4. logger.error
5. logger.log

## Test

```shell
npm run test
```

## Functions

1. Support user define the slice of log file size and log file path.

2. Only logger.debug doesn't log to file, all of them can be print in console.

3. Only identify 999 levels in input value which contains object or array.

4. Support to print part of req and res(big object) in nodejs

5. Data type includes Number, String, Undefined, Null, Boolean, Object, Array, Function, Error, Set, Map and Symbol can be support stringify. If something can't be stringify, it will be print data type, such as DOM element, it will print '[object HTMLDivElement]'

6. Although we support print Function, Set, Map and Symbol directly, we still don't recommend to print it without any transfer, and it will give a warn to remind you if you set dataTypeWarn to be true.

7. Enable productionModel will not print log to console, it will be fast for program in production environment.

8. In default, it will log file by log level and it will log into multiple different kinds of files, if you specify logFilePath as a path string, beauty-logger will print all logs to one file.

9. Support new multiple beauty-logger instance to log different files.

10. All of logger method support Promise chain call

11. Reserve data type which is in console, only stringify data which will be written to file.

12. Only support logger.debug, logger.info, logger.warn, logger.error and logger.log

13. Support nodejs at least 6.x

14. support more beauty logger working at the same time and print which file execute it correctly.

15. execute callback when logging

## Preview pictures

[![log_example_1](https://github.com/zhoushoujian/beauty-logger/blob/master/docs/log_example_1.png)](https://github.com/zhoushoujian/beauty-logger/blob/master/docs/log_example_1.png)

## License

[MIT](https://github.com/zhoushoujian/beauty-logger/blob/master/LICENSE)


CHANGELOG.md
# CHANGELOG

## 4.7.4 (2021.12.21)

1. 在 windows 系统上自动隐藏 cmd 窗口

## 4.7.3 (2021.12.08)

1. 不限制日志文件的分片数量
2. 日志系统报错记录到日志

## 4.7.2 (2021.12.01)

1. 修复数据上报

## 4.7.1 (2021.11.26)

1. 新增 executeCommand 静态方法

## 4.7.0 (2021.09.27)

1. 日志格式中新增主机名

## 4.6.0 (2021.09.04)

1. 新增包使用情况上报

## 4.5.0 (2021.08.26)

1. 适配 console-format 2.0.0 版本

## 4.4.0 (2021.08.09)

1. 日志回调函数增加纯净的日志参数
2. 完善 ts 类型推导

## 4.3.1 (2021.07.29)

1. 修复打印 Date 对象
2. 修复递归打印 Set 和 Map 数据类型

## 4.3.0 (2021.07.09)

新增打印日志时并行的回调任务

## 4.2.1 (2021.07.07)

原生支持多实例合并打印并正确显示调用的文件

## 4.1.0 (2021.07.03)

修复不能日志文件分片的 bug  
修复某些日志文件存放路径不正确的问题

## 4.0.2 (2021.06.03)

使用 ts 重构

## 4.0.0 (2021.06.03)

打印执行文件所在的路径

## 3.0.1 (2021.05.29)

打印信息中添加进程号

## 3.0.0 (2021.03.29)

移除参数 enableMultipleLogFile，根据 logFilePath 的数据类型来判断打印到多文件还是单文件

## 2.1.1 (2021.01.10)

调整 module.exports 和 define 的导出顺序,优先导出 module.exports

## 2.1.0 (2021.01.05)

完善测试脚本

## 2.0.9 (2020.12.13)

增加 log 方法


beauty-logger.ts
const consoleFormat = require('console-format');

consoleFormat();

declare namespace NodeJS {
  interface Global {
    beautyLogger: IBeautyLoggerInstance;
  }
}

interface IBeautyLoggerInstance {
  currentProjectPath: string;
  userConfig: IUserConfig[] | ILogQueue[] | IUserConfig;
  debug: (a: any[]) => void;
  info: (a: any[]) => void;
  warn: (a: any[]) => void;
  error: (a: any[]) => void;
  log: (a: any[]) => void;
}

interface IUserConfig {
  beautyLogger: boolean;
  loggerFilePath: string;
  logFileSize: number;
  dataTypeWarn: boolean;
  productionModel: boolean;
  onlyPrintInConsole: boolean;
  otherBeautyLoggerInstances: IBeautyLoggerInstance;
  callback: (level: ILevel, data: string, pid: number, filePath: string, content: string) => void;
}

type ILevel = 'debug' | 'info' | 'warn' | 'error' | 'log';

interface ILogQueue {
  level: ILevel;
  buffer: string;
}

const LOGGER_LEVEL = ['debug', 'info', 'warn', 'error', 'log'],
  isNodeJs = typeof process === 'object';

let loopTimes = 0,
  fs: any,
  path: any,
  deepcopy: any;

function dealWithItems(item: any, needWarn: boolean) {
  try {
    const dist = deepcopy(item);
    return JSON.stringify(
      dist,
      function (_key, value) {
        loopTimes = 0;
        return formatDataType(value, needWarn);
      },
      0,
    );
  } catch (err) {
    return Object.prototype.toString.call(item);
  }
}

function formatDataType(value: any, needWarn: boolean) {
  loopTimes++;
  let formattedOnes: any = '';
  try {
    const valueType = Object.prototype.toString.call(value);
    switch (valueType) {
      case '[object Number]':
      case '[object String]':
      case '[object Undefined]':
      case '[object Null]':
      case '[object Boolean]':
        formattedOnes = value;
        break;
      case '[object Object]':
      case '[object Array]':
        for (const i in value) {
          if (Object.prototype.hasOwnProperty.call(value, i)) {
            if (loopTimes > 999) {
              value[i] = Object.prototype.toString.call(value[i]);
            } else {
              value[i] = formatDataType(value[i], needWarn);
            }
          }
        }
        formattedOnes = value;
        break;
      case '[object Function]':
        if (needWarn) {
          console.warn("we don't recommend to print function directly", value);
        }
        formattedOnes = Function.prototype.toString.call(value);
        break;
      case '[object Error]':
        formattedOnes = value.stack || value.toString();
        break;
      case '[object Symbol]':
      case '[object Date]':
        if (needWarn) {
          console.warn("we don't recommend to print Symbol directly", value);
        }
        formattedOnes = value.toString();
        break;
      case '[object Set]':
        if (needWarn) {
          console.warn("we don't recommend to print Set directly", value);
        }
        formattedOnes = formatDataType(Array.from(value), needWarn);
        break;
      case '[object Map]': {
        if (needWarn) {
          console.warn("we don't recommend to print Map directly", value);
        }
        const obj: any = {};
        value.forEach(function (item: any, key: any) {
          obj[key] = item;
        });
        formattedOnes = formatDataType(obj, needWarn);
        break;
      }
      default:
        formattedOnes = Object.prototype.toString.call(value);
        break;
    }
  } catch (err) {
    formattedOnes = {};
  }
  return formattedOnes;
}

function getLogPath(level: ILevel) {
  const enableMultipleLogFile = this.userConfig.enableMultipleLogFile;
  const loggerFilePath = this.userConfig.loggerFilePath;
  if (enableMultipleLogFile) {
    return loggerFilePath[level];
  } else {
    return loggerFilePath;
  }
}

function logInFile(buffer: string, level: ILevel) {
  return checkFileState.bind(this)(level).then(writeFile.bind(this, buffer, level));
}

function checkFileState(level: ILevel) {
  // check file existed and file size
  const self = this;
  return new Promise(function (resolve) {
    const logFileSize = self.userConfig.logFileSize;
    const currentProjectLoggerFolder = self.userConfig.currentProjectLoggerFolder;
    if (!fs.existsSync(getLogPath.bind(self)(level))) {
      fs.appendFileSync(getLogPath.bind(self)(level), '');
      return resolve(0);
    } else {
      return fs.stat(getLogPath.bind(self)(level), function (err: Error, stats: { size: number }) {
        if (err) {
          console.debug('beauty-logger: checkFileState fs.stat err', err);
          return resolve(0);
        } else {
          // logger is async, so one logger has appendFile after next one check file state
          if (stats && stats.size > logFileSize) {
            fs.readdir(currentProjectLoggerFolder, function (err: Error, files: string[]) {
              if (err) {
                console.debug('beauty-logger: checkFileState fs.stat fs.readdir err', err);
              }
              const currentLogFilename = path.parse(getLogPath.bind(self)(level)).name;
              const currentLogFileExtname = path.parse(getLogPath.bind(self)(level)).ext;
              let currentLogFileExtnameWithoutDot = '';
              if (currentLogFileExtname) {
                currentLogFileExtnameWithoutDot = currentLogFileExtname.replace('.', '');
              }
              const fileList = files.filter(function (file) {
                return RegExp('^' + currentLogFilename + '[0-9]*.*' + currentLogFileExtnameWithoutDot + '*$').test(
                  file,
                );
              });
              for (let i = fileList.length; i > 0; i--) {
                if (i >= 10) {
                  fs.unlinkSync(currentProjectLoggerFolder + '/' + fileList[i - 1]);
                  continue;
                }
                fs.renameSync(
                  currentProjectLoggerFolder + '/' + fileList[i - 1],
                  currentProjectLoggerFolder + '/' + currentLogFilename + i + currentLogFileExtname,
                );
              }
              resolve(0);
            });
          } else {
            return resolve(0);
          }
        }
      });
    }
  });
}

function writeFile(buffer: string, level: ILevel) {
  const self = this;
  return new Promise(function (res) {
    fs.writeFile(
      getLogPath.bind(self)(level),
      buffer,
      {
        flag: 'a+',
      },
      function (err: Error) {
        if (err) console.debug('beauty-logger: writeFile err', err.stack || err.toString());
        self.logQueue.shift();
        res(buffer);
        if (self.logQueue.length) {
          const firstItem = self.logQueue[0];
          return logInFile.call(self, firstItem.buffer, firstItem.level);
        }
      },
    );
  });
}

function InitLogger(config = {} as IUserConfig) {
  if (Object.prototype.toString.call(config) === '[object Object]') {
    if (isNodeJs) {
      try {
        fs = require('fs');
        path = require('path');
        deepcopy = require('deepcopy');
        this.logQueue = [] as ILogQueue[];
        this.userConfig = config;
        const currentProjectPath = process.cwd().split('node_modules')[0];
        this.userConfig.loggerFilePath = {
          info: currentProjectPath + '/INFO.log',
          warn: currentProjectPath + '/WARN.log',
          error: currentProjectPath + '/ERROR.log',
          log: currentProjectPath + '/LOG.log',
        };
        this.userConfig.beautyLogger = true;
        this.userConfig.logFileSize =
          typeof this.userConfig.logFileSize === 'number' ? this.userConfig.logFileSize : 1024 * 1024 * 10;
        this.userConfig.dataTypeWarn =
          typeof this.userConfig.dataTypeWarn === 'boolean' ? this.userConfig.dataTypeWarn : false;
        this.userConfig.productionModel =
          typeof this.userConfig.productionModel === 'boolean' ? this.userConfig.productionModel : false;
        this.userConfig.onlyPrintInConsole =
          typeof this.userConfig.onlyPrintInConsole === 'boolean' ? this.userConfig.onlyPrintInConsole : false;
        this.userConfig.otherBeautyLoggerInstances = Array.isArray(this.userConfig.otherBeautyLoggerInstances)
          ? this.userConfig.otherBeautyLoggerInstances
          : [];
        this.userConfig.callback = this.userConfig.callback instanceof Function ? this.userConfig.callback : null;
        let levelArr = ['info', 'warn', 'error', 'log'];
        if (Object.prototype.toString.call(this.userConfig.logFilePath) === '[object Object]') {
          for (const i in this.userConfig.logFilePath) {
            if (Object.prototype.hasOwnProperty.call(this.userConfig.logFilePath, i) && i !== undefined) {
              levelArr = levelArr.filter(item => item !== i.toLocaleLowerCase());
              this.userConfig.loggerFilePath[i] = this.userConfig.logFilePath[i];
            }
          }
          this.userConfig.enableMultipleLogFile = true;
        } else if (typeof this.userConfig.logFilePath === 'string') {
          this.userConfig.loggerFilePath = this.userConfig.logFilePath;
          this.userConfig.enableMultipleLogFile = false;
        } else if (typeof this.userConfig.logFilePath === 'undefined') {
          //use default value
        } else {
          throw new Error('beauty-logger: logFilePath must be an object or empty');
        }
        this.userConfig.currentProjectLoggerFolder = path.parse(getLogPath.bind(this)('info')).dir;
        if (this.userConfig.enableMultipleLogFile) {
          levelArr.forEach(item => {
            this.userConfig.loggerFilePath[item] =
              this.userConfig.currentProjectLoggerFolder + '/' + item.toUpperCase() + '.log';
          });
        }
        if (!global.beautyLogger) {
          global.beautyLogger = {} as IBeautyLoggerInstance;
          global.beautyLogger.currentProjectPath = currentProjectPath;
          global.beautyLogger.userConfig = [];
        }
        (global.beautyLogger.userConfig as IUserConfig[]).push(this.userConfig);
        (global.beautyLogger.userConfig as IUserConfig[]).push(this.logQueue);
        if (typeof this.userConfig.loggerFilePath === 'string') {
          if (!fs.existsSync(this.userConfig.loggerFilePath)) {
            fs.appendFileSync(this.userConfig.loggerFilePath, '');
          }
        } else {
          for (const i in this.userConfig.loggerFilePath) {
            if (Object.prototype.hasOwnProperty.call(this.userConfig.loggerFilePath, i)) {
              if (!fs.existsSync(this.userConfig.loggerFilePath[i])) {
                fs.appendFileSync(this.userConfig.loggerFilePath[i], '');
              }
            }
          }
        }
      } catch (err) {
        console.error('beauty-logger: err', err);
      }
    }
  } else {
    throw new Error('beauty-logger: config must be an object or empty');
  }
}

function loggerInFile(level: ILevel, data = '') {
  if (isNodeJs) {
    const productionModel = this.userConfig.productionModel;
    const dataTypeWarn = this.userConfig.dataTypeWarn;
    const onlyPrintInConsole = this.userConfig.onlyPrintInConsole;
    if (!productionModel) {
      //@ts-ignore
      console[level].apply(null, Array.prototype.slice.call(arguments).slice(1));
    }
    if (level === 'debug' || onlyPrintInConsole) return;
    loopTimes = 0;
    let dist = deepcopy(data);
    dist = JSON.stringify(
      dist,
      function (_key, value) {
        return formatDataType(value, dataTypeWarn);
      },
      0,
    );
    let extend: string[] | string = [];
    const args = Array.prototype.slice.call(arguments).slice(2);
    if (args.length) {
      extend = args.map(function (item) {
        return dealWithItems(item, dataTypeWarn);
      });
      if (extend.length) {
        extend = `  [ext] ${extend.join('')}`;
      }
    }
    const content = `${dist} ${extend} \r\n`;
    const filePath = consoleFormat.dealWithFilePath();
    const buffer = `[${consoleFormat.getTime()}] [${level.toUpperCase()}] [${process.pid}] [${filePath}] ${content}`;
    if (Object.prototype.toString.call(this.userConfig.callback) === '[object Function]') {
      this.userConfig.callback(level, buffer, process.pid, filePath, content);
    }
    if (this.logQueue.length) {
      this.logQueue.push({
        level,
        buffer,
      });
      return Promise.resolve(buffer);
    }
    this.logQueue.push({
      level,
      buffer,
    });
    return logInFile.call(this, buffer, level);
  } else {
    //@ts-ignore
    console[level].apply(null, Array.prototype.slice.call(arguments).slice(1));
  }
}

(LOGGER_LEVEL as ILevel[]).forEach(function (level: ILevel) {
  InitLogger.prototype[level] = function (data: any, ...args: any) {
    if (this && this.userConfig) {
      if (this.userConfig.otherBeautyLoggerInstances && this.userConfig.otherBeautyLoggerInstances.length) {
        this.userConfig.otherBeautyLoggerInstances.forEach((item: IBeautyLoggerInstance) => {
          if (item && item.userConfig) {
            if ((item.userConfig as IUserConfig).beautyLogger) {
              item[level]([data, ...args]);
            }
          }
        });
      }
    }
    //@ts-ignore
    return loggerInFile.bind(this)(level, data, ...args);
  };
});

// Export to popular environments boilerplate.
if (typeof module !== 'undefined' && module.exports) {
  module.exports = InitLogger;
  //@ts-ignore
  // eslint-disable-next-line no-undef
} else if (typeof define === 'function' && define.amd) {
  //@ts-ignore
  // eslint-disable-next-line no-undef
  define(InitLogger);
} else {
  InitLogger._prevLogger = this.InitLogger;

  InitLogger.noConflict = function () {
    //@ts-ignore
    this.InitLogger = InitLogger._prevLogger;
    return InitLogger;
  };
  this.InitLogger = InitLogger;
}
